### PR TITLE
Handle special learning rates better

### PR DESCRIPTION
**Pull request by albertfgu**
This PR makes a pretty large change to the way special hyperparameters are handled. Instead of being controlled by the global training loop (`train.py` has to specify the special parameters and their hyperparameters), it's better to have the models locally handle their own hyperparameters.

This required several changes to make. The idea is that each Flax nn.Module should just store a dictionary of its hyperparameters. For example, `S4Layer` defines
```
lr = {
    "Lambda_re": 0.1,
    "Lambda_im": 0.1,
    "P": 0.1,
    "B": 0.1,
    "log_step": 0.1,
}
```
which says that it wants each of these hyperparameters to have LR scaled by 0.1. (All special parameters also have weight decay set to 0). The optimizer constructor simply has to grab `model_cls.lr` from the core Module layer to get access to these.

--------------------------

However, these Modules were actually previously handled by another layer of indirection, through the `SSLayerInit` functions which passed things like `N` and initializations into a `partial` call, e.g.
```
def SSMInit(N):
    return partial(cloneLayer(SSMLayer), A=make_HiPPO(N), N=N)
```
Thus the actual layer constructors were not the original nn.Module, but wrapped `partial` calls that lost access to the `lr` attribute.

--------------------------

In general this design seems suboptimal, so the first change is to get rid of these partial constructors. Instead, the `N` value has to be passed directly into the layer. This leads to another problem: the layer is wrapped inside several other modules e.g. (`SequenceBlock`, `StackedModule`, and `BatchStackedModule`) which must handle all the potential options of the inside layer. Instead, I gave them a `layer_args` parameter to pass directly into the core layer constructor. This would hold things like `N`, `l_max`, and anything else that only the layer needs.

--------------------------

A second problem is that the vanilla `SSMLayer` was defined with the HiPPO A matrix, as shown above. This was sort of a special case that's annoying to handle, as we don't want to construct `layer_args["A"] = make_hippo()` only for this class. Thus I decided to just get rid of the HiPPO init from the vanilla SSMLayer entirely and randomly initialize A. This required changing the blog post, because it defined HiPPO first then this `SSMLayer`.

So I defined `SSMLayer` as a generic SSM and moved the HiPPO definition into a Part 1b afterwards. Actually, I think this is cleaner from the perspective of the S4 story as well. It makes sense to define SSMs in general and then talk about how S4 is different (modeling side: use HiPPO; computational side: use DPLR)

--------------------------

- Finally, there was a bug in the original code where the LR schedulers weren't handled correctly for the special parameters. The new code takes care of all of this in a cleaner way.
- We can also add trainable parameter counts using the same ideas

### CODE REVIEW

**Review by srush — APPROVED** [review]
I really like this change, and the removal of the indirection. My only comment is that it would be nice if it failed if you spelled an argument name wrong. I think we could add a check that the dictionary exists as a variable name. 

Can I understand what the bug was as well, since we may need to fix that in HF version as well. Is it that the S4 extra parameters should be 0.1 multiplicative of the learning rate always?

**Comment by albertfgu** [comment]
Good idea on the check: I'll add that later (too annoying to rebase things around to put it in this PR)

Right, the bug is that LR flags should specify the max LR and the rest of the schedule should have the same shape. 

Note: Actually in the original S4 code we had special LRs set to min(1e-3, lr) but here I use a constant multiplier. It's easy to change either way with this new setup. It might make a slight difference sometimes but hopefully not.
